modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java

```diff
@@ -705,4 +706,56 @@ public void stopMessageExpirationWorker() {
     public SlotCoordinator getSlotCoordinator() {
         return slotCoordinator;
     }
+
+    /**
+     * Store retained messages in the message store.
+     *
+     * @see org.wso2.andes.kernel.AndesMessageMetadata#retain
+     * @param retainList
+     *         Retained message list
+     */
+    public void storeRetainedMessages(List<AndesMessage> retainList) throws AndesException {
+        messageStore.storeRetainedMessages(retainList);
+    }
+
+
+    /**
+     * Return matching retained message metadata for the given subscription topic name. An empty list is returned if no
+     * match is found.
+     *
+     * @param subscriptionTopicName
+     *         Destination string provided by the subscriber
+     * @return AndesMessageMetadata
+     * @throws AndesException
+     */
+    public List<AndesMessageMetadata> getRetainedMessageByTopic(String subscriptionTopicName) throws AndesException {
+        List<AndesMessageMetadata> retainMessageList = new ArrayList<AndesMessageMetadata>();
+        List<String> topicList = messageStore.getAllRetainedTopics();
+
+        for (String topicName : topicList) {
+            if (TopicParserUtil.isMatching(topicName, subscriptionTopicName)) {
+                retainMessageList.add(messageStore.getRetainedMetaData(topicName));
+            }
+        }
+
+        return retainMessageList;
+    }
+
+    /**
+     * Return message content for the given retained message metadata.
+     *
+     * @param metadata
+     *         Message metadata
+     * @return AndesContent
+     * @throws AndesException
+     */
+    public AndesContent getRetainedMessageContent(AndesMessageMetadata metadata) throws AndesException {
+        long messageID = metadata.getMessageID();
+        int contentSize = AMQPUtils.convertAndesMetadataToAMQMetadata(metadata).getContentSize();
```

Comment:
the name 'convertAndesMetadataToAMQMetadata' is odd and AMQPUtils given that it is called from messagingengine. Shall we have a look?